### PR TITLE
feat: Improve console output to show packet enum names (magic_enum)

### PR DESCRIPTION
**Pull request by jadebenn**
Replaces #1340. Adds magic_enum as a dependency in order to enable enum reflection and display world and packet message enum strings in the server output.

### CODE REVIEW

**Comment by jadebenn** [comment]
Marking as draft because I had an epiphany about how to make the performance of our magic_enum implementation equal or exceed that of the preprocessor implementation. I will test this later, but here's the idea:

Instead of directly using the enum_name() function magic_enums provides, I will create a wrapper function called "StringifiedEnum" that defines a static const std::array provided by magic_enum's similarly-named enum_names() function. This function will then pass a reference to the corresponding array entry to the calling function. So, this should load a single copy of the pre-sorted(?) std::array into memory at compile-time, and it will be templated-out so those arrays are only generated for the enums this function is called on.

This should essentially eliminate any and all performance overhead magic_enums could possibly impose, as after compilation only an array of entries corresponding to the enum would exist, and a function to access it.

**Comment by jadebenn** [comment]
Implemented more or less what I said I would. The desired performance has been achieved, but I'd appreciate a look at the iteration algorithm I'm using, which was borrowed from @EmosewaMC's work on the other branch. One issue I'm having in particular is that it doesn't handle "gaps" in the enum by returning a NULL value, and I don't currently understand arrays well enough to fix that.

**Comment by jadebenn** [comment]
I think this is as good as it gets now. Much better performance than using the magic_enum function directly and restored the ability to see "gaps" in the enum. Only thing I haven't tested is if thhere are any enums out of order...

**Comment by jadebenn** [comment]
Actually, we're good. Magic_enum gives us a sorted array.

File: dWorldServer/WorldServer.cpp
Comment (EmosewaMC) [inline dWorldServer/WorldServer.cpp:1248]: This is already bounds checked far above in the same function if it gets this low

File: dWorldServer/WorldServer.cpp
Comment (EmosewaMC) [inline dWorldServer/WorldServer.cpp:1250]: This should be able to just be reinterpret_cast straight to eWorldMessageType, can it not? You could also inline these two to be `const char* messageIdString = StringifiedEnum::ToString(*reinterpret_cast<eWorldMessageType*>(&packet->data[3])).data();`

File: tests/dCommonTests/dEnumsTests/CMakeLists.txt
Comment (EmosewaMC) [inline tests/dCommonTests/dEnumsTests/CMakeLists.txt:10]: We don't write these like this right now, so for consistency this should follow how the other folder names are gotten (CMakeLists.txt in this directory just globs its files and sends to PARENT_SCOPE, then the parent scope prepends the folder name it just added)

File: tests/dCommonTests/dEnumsTests/MagicEnumTests.cpp
Comment (EmosewaMC) [inline tests/dCommonTests/dEnumsTests/MagicEnumTests.cpp:82]: These can be combined with the above macro and then you can either use templates to plug in the tested type, or a define which takes an argument of the string to test to reduce repeated code

Comment (EmosewaMC) [reply tests/dCommonTests/dEnumsTests/MagicEnumTests.cpp:82]: The final call in log_test_invalid can also be simplified to a .empty call with no need to compare with NULL.  What do you think of that change?

File: tests/dCommonTests/dEnumsTests/MagicEnumTests.cpp
Comment (EmosewaMC) [inline tests/dCommonTests/dEnumsTests/MagicEnumTests.cpp:116]: I would recommend since I didn't do this at the moment, using srand(time(NULL)) to at least seed the rng to the current epoch

File: CMakeLists.txt
Comment (EmosewaMC) [inline CMakeLists.txt:295]: if we are vendoring magic_enum, we should be using their cmake instead of just including their directories as we do for the other third parties

**Review by EmosewaMC — COMMENTED** [review]
left some comments that should be talked about; also is there anything more to adding a submodule others need to be aware of when updating dlu or is the README already taking care of this?

File: tests/dCommonTests/dEnumsTests/CMakeLists.txt
Comment (aronwk-aaron) [reply tests/dCommonTests/dEnumsTests/CMakeLists.txt:10]: Currently, the tests are different and other tests cmakeslists are like this.
Leave it as is and we can make them consistent in a separate PR

File: tests/dCommonTests/dEnumsTests/MagicEnumTests.cpp
Comment (EmosewaMC) [inline tests/dCommonTests/dEnumsTests/MagicEnumTests.cpp:11]: nit: macro should be caps locked

File: tests/dCommonTests/dEnumsTests/MagicEnumTests.cpp
Comment (EmosewaMC) [inline tests/dCommonTests/dEnumsTests/MagicEnumTests.cpp:105]: I also noticed the timing code is repeated, could that be simplified down to one call as well?

File: tests/dCommonTests/dEnumsTests/MagicEnumTests.cpp
Comment (EmosewaMC) [inline tests/dCommonTests/dEnumsTests/MagicEnumTests.cpp:13]: You dont need a .data call here I dont believe?  You can just compare the stringview directly to a const char*?

**Comment by jadebenn** [comment]
That's it for me. Ready to merge this when you all are.

**Review by aronwk-aaron — APPROVED** [review]

File: dCommon/dEnums/StringifiedEnum.h
Comment (EmosewaMC) [inline dCommon/dEnums/StringifiedEnum.h:11]: nit: default initializer unnecessary here.  std containers are already default initialized

File: dCommon/dEnums/StringifiedEnum.h
Comment (EmosewaMC) [inline dCommon/dEnums/StringifiedEnum.h:22]: if these are going to be new lined, then we might as well use braces for clarity

File: tests/dCommonTests/dEnumsTests/MagicEnumTests.cpp
Comment (EmosewaMC) [inline tests/dCommonTests/dEnumsTests/MagicEnumTests.cpp:126]: Is there a reason this is a macro and not a function?

Comment (EmosewaMC) [reply tests/dCommonTests/dEnumsTests/MagicEnumTests.cpp:126]: Is there also a reason its always logging the first argument as wmArray, when it should be using the first param of LOG_EARRAY as EARRAY_VAR?

**Review by EmosewaMC — COMMENTED** [review]
more feedback

File: tests/dCommonTests/dEnumsTests/MagicEnumTests.cpp
Comment (jadebenn) [reply tests/dCommonTests/dEnumsTests/MagicEnumTests.cpp:126]: Nope, just a typo

Comment (jadebenn) [reply tests/dCommonTests/dEnumsTests/MagicEnumTests.cpp:126]: As for it being a function: It's a real pain in the behind to deal with the array type as a function argument and this code will only be used here, so...

**Review by aronwk-aaron — APPROVED** [review]

**Review by EmosewaMC — APPROVED** [review]
Some odd spacing on the edited else but im not going to block on it.  Would recommend if you don't already alt+shift+f to format a file to the repo standard if it currently doesn't meet it (for vscode on windows at least this hotkey works)

**Pull request merged**
